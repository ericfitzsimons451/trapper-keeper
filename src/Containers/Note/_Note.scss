.note {
    display: inline-block;
    text-align: left;
    position: relative;
    padding: 5px;
    width: 200px;
    border: 1px solid #e0e0e0;
    border-radius: 5px;
    margin:5px;
    .delete-button {
        cursor: pointer;
        border: none;
        background: none;
        top: -5px;
        right: -5px;
        outline: none;
        height: 20px;
        width: 20px;
        background-image: url("./images/delete.svg");
        background-size: 20px 20px;        
        position: absolute;
        transition: 0.2s;
        opacity: 0;
    }
    &:hover{
        box-shadow: 2px 2px 5px #e0e0e0;
        .delete-button {
            opacity: 1;
        } 
        cursor: pointer;
    }
    
    p{
        margin: 0px 0px 0px 10px;
        text-wrap: wrap;
    }
    h2 {
        margin: 10px 0px 0px 15px;
        overflow: hidden;
        text-overflow: ellipsis;
    }
}

.check-section {
    border-top: 1px solid #e0e0e0;
}

.check-text {
   text-decoration: line-through; 
}