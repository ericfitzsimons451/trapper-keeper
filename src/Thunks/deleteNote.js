import { deleteCard, isLoading, hasError } from "../actions";

export const deleteNote = (id) => {
  return async (dispatch) => {
    try {
      dispatch(isLoading(true));
      const response = await fetch(`http://localhost:3000/api/v1/notes/${id}`, {
        method: "DELETE"
      });
      if (response.status !== 204) {
        throw new Error(response.statusText);
      }
      dispatch(isLoading(false));
      dispatch(deleteCard(id));
    } catch (error) {
      console.log("hello world")
      dispatch(hasError(error.message));
    }
  };
};
