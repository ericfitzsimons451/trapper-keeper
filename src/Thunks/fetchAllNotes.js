import { getAllNotes, hasError, isLoading } from '../actions'

export const fetchAllNotes = (url) => {
  return async (dispatch) => {
    try {
      dispatch(isLoading(true))
      const response = await fetch(url)
      if (!response.ok) {
        throw new Error(response.statusText)
      }
      const notes = await response.json()
      dispatch(isLoading(false))
      dispatch(getAllNotes(notes))
    } catch (error) {
      dispatch(hasError(error.message))
    }
  }
}