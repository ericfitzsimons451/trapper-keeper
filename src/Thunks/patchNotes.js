import { hasError } from '../actions'

export const patchNotes = (notes) => {
	const option = {
		method: "PATCH",
		headers: {
			'Content-Type': 'application/json'
		},
		body: JSON.stringify(notes)
	}

	return async (dispatch) => {
		try {
			const response = await fetch("http://localhost:3000/api/v1/notes", option)
			if (!response.ok) {
				throw new Error(response.statusText)
			}
		} catch (error) {
			dispatch(hasError(error.message))
		}
	}
}