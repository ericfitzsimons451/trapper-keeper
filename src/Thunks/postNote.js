import { addNote, isLoading, hasError } from '../actions'

export const postNote = (note) => {
    return async (dispatch) => {
        try {
            dispatch(isLoading(true))
            const response = await fetch('http://localhost:3000/api/v1/notes')
            if (!response.ok) {
                throw new Error(response.statusText)
            }
            const note = await response.json()
            dispatch(isLoading(false))
            dispatch(addNote(note))
        } catch (error) {
            dispatch(hasError(error.message))
        }
    }
}