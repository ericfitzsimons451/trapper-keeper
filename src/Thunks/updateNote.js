import { updateNote, isLoading, hasError } from '../actions'

export const editNote = (noteToUpdate) => {
  console.log('Update Note Thunk Fires')
  const option = {
    method: 'PUT',
    body: JSON.stringify(noteToUpdate),
    headers: {
      'Content-Type': 'application/json'
    }
  }
  return async (dispatch) => {
    try {
      dispatch(isLoading(true))
      const response = await fetch(`http://localhost:3000/api/v1/notes/${noteToUpdate.id}`, option)
      if (!response.ok) {
        throw new Error(response.statusText)
      }
      const note = await response.json()
      dispatch(isLoading(false))
      console.log(noteToUpdate)
      dispatch(updateNote(noteToUpdate))
    } catch (error) {
      dispatch(hasError(error.message))
    }
  }
}