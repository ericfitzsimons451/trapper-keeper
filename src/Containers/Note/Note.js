import React, { Component } from "react"
import { deleteNote } from "../../Thunks/deleteNote";
import { editNote } from "../../Thunks/updateNote";
import { startDrag } from "../../actions"
import { changeNoteOrder } from "../../actions"
import { patchNotes } from "../../Thunks/patchNotes"
import { connect } from "react-redux"

export class Note extends Component{
    // refactor to take out constructor
    constructor(){
        super();
        this.state = {
            listItems: []
        }
    }

    componentDidMount = () => {
        this.setState({
            listItems: this.props.note.listItems,
        })
    }

    toggleCheckBox = e => {
        const { id } = e.target.parentElement;
        const newBody = this.state.listItems.map(note => {
          if (note.id === parseInt(id)) {
            return { text: note.text, checked: !note.checked, id: note.id };
          }
          return note;
        });
        this.setState({
          listItems: newBody
        }, this.editNoteCheck);
      };

      editNoteCheck = async () => {
        const updatedNote = {
            id: this.props.note.id,
            title: this.props.note.title,
            listItems: this.state.listItems
          };
          await this.props.editNote(updatedNote);
      }

    openNote = e => {
        if(e.target.className !== "checkbox" && e.target.className !== "uncheckbox" && e.target.className !== "delete-button" ){
            this.props.history.push(`/notes/${this.props.note.id}`)
        }
      }

    onDragOver = (id) => {
        if(this.props.startID === id){
            return
        }else{
            this.props.changeNoteOrder(this.props.startID, id)
        }
     }

    onDragEnd = () => {
        this.props.patchNotes(this.props.notes)
    }

    render = () => {
        const filteredUnChecked = this.state.listItems.filter(note => !note.checked);
        const filteredChecked = this.state.listItems.filter(note => note.checked);
        const unchecked = filteredUnChecked.map(text => {
            return (
              <div key={text.id} id={text.id} className="text">
                <div onClick={this.toggleCheckBox} className="uncheckbox" />
                <p>{text.text}</p>
              </div>
            );
          });
          const checked = filteredChecked.map(text => {
            return (
              <div key={text.id} id={text.id} className="text">
                <div onClick={this.toggleCheckBox} className="checkbox" />
                <p className="check-text">{text.text}</p>
                </div>
            );
          });
        return(
            <div draggable onDragStart={() => {this.props.startDrag(this.props.note.id)}} onDragOver={() => {this.onDragOver(this.props.note.id)}} onDragEnd={this.onDragEnd} className="note" onClick={this.openNote}>
                <button className="delete-button"
                  onClick={() => {
                    this.props.deleteNote(this.props.note.id);
                  }}
                ></button>
                <h2>{this.props.note.title}</h2>
                {
                    unchecked
                }
                {
                    checked.length > 0 && (
                        <div className="check-section">
                            {
                                checked
                            }
                        </div>
                    )
                }
            </div>
        )
    }
}

export const mapDispatchToProps = (dispatch) => ({
    deleteNote: note => dispatch(deleteNote(note)),
    editNote: note => dispatch(editNote(note)),
    startDrag: id => dispatch(startDrag(id)),
    changeNoteOrder: (noteOne, noteTwo) => dispatch(changeNoteOrder(noteOne, noteTwo)),
    patchNotes: notes => dispatch(patchNotes(notes))
})

export const mapStateToProps = (state) => ({
    startID: state.startID,
    notes: state.notes
})

export default connect(mapStateToProps, mapDispatchToProps)(Note)