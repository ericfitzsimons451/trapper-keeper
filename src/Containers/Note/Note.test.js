
import React from 'react'
import { Note } from './Note'
import { mapDispatchToProps, mapStateToProps } from './Note'
import { shallow } from 'enzyme'
import { deleteNote } from '../../Thunks/deleteNote'
jest.mock('../../Thunks/deleteNote')

describe('Note', () => {
	let wrapper;
	let mockNote;
	let mockState;
	let mockHistory;

	beforeEach(() => {
		mockState = [
			{ id: 4, text: 'yo', checked: false }, 
			{ id: 5, text: 'hello', checked: false }
		]
		mockHistory = { push: jest.fn() }
		mockNote = {
			title: 'hello',
			listItems: [
				{ id: 4, text: 'yo', checked: false }, 
				{ id: 5, text: 'hello', checked: false }
			]
		}
		wrapper = shallow(
		<Note
			note={mockNote}
			deleteNote={jest.fn()}
			editNote={jest.fn()}
			startDrag={jest.fn()}
			changeNoteOrder={jest.fn()}
			history={mockHistory} />)
	})

	it('should match the snapshot', () => {
		expect(wrapper).toMatchSnapshot()
	})

	it('should have initial state', () => {
		expect(wrapper.state()).toEqual({ listItems: mockState })
	})

	it('should mapStateToProps', () => {
		mockState = { startID: 29 }
		const results = mapStateToProps(mockState)
		expect(results).toEqual(mockState)
	})

	it('should mapDispatchToProps', () => {
		const mockDispatch = jest.fn()
		const mockNote = [
			{
				id: 1242, title: "test",
				body:
					[{
						context: "testone",
						isChecked: false
					}, {
						context: "testtwo",
						isChecked: false
					}, {
						context: "testthree",
						isChecked: true
					}]
			}
		]
		const actionToDispatch = deleteNote(mockNote)
		const mappedProps = mapDispatchToProps(mockDispatch)
		mappedProps.deleteNote(mockNote)
		expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
	})

	describe('componentDidMount', () => {
		it('should call setState inside of componentDidMount', () => {
			const mockListItems = [{}]
			wrapper.setState({ listItems: mockListItems })
			expect(wrapper.state()).toEqual({ listItems: mockListItems })
		})
	})

	describe('toggleCheckBox', () => {
		it('should reset state with the updatedNote if the ID matches', () => {
			const mockEvent = { target: { parentElement: { id: 4 } } }
			wrapper.instance().toggleCheckBox(mockEvent)
			expect(wrapper.state()).toEqual({
				listItems: [{ id: 4, text: 'yo', checked: true }, { id: 5, text: 'hello', checked: false }]
			})
		})

		// it('should fire editNoteCheck as a callback function', async () => {
		//     const mockEvent = {target: {parentElement: {id: 4 }}}
		//     const editNoteCheck = jest.fn()
		//     await wrapper.instance().toggleCheckBox(mockEvent)
		//     expect(editNoteCheck).toHaveBeenCalled()
		// })
	})

	describe('editNoteCheck', () => {
		it('should fire props.editNote', async () => {
			const mockNote = {
				title: 'hello',
				listItems: mockState
			}
			await wrapper.instance().editNoteCheck()
			expect(wrapper.instance().props.editNote).toHaveBeenCalledWith(mockNote)
		})
	})

	describe.skip('openNote', () => {

		// these tests need some work.  I am not explicitly calling
		// the push onto the history object.  Ideas?


		it('should push a note id onto the history object if the className is NOT checkbox, uncheckbox, or delete-button', () => {
			const mockPushToHistory = jest.fn()
			const mockEvent = { target: { parentElement: { className: 'update' } } }
			wrapper.instance().openNote(mockEvent)
			expect(mockPushToHistory).toHaveBeenCalled()
		})

		it('should call nothing if the className is CHECKBOX, UNCHECKBOX, OR DELETE-BUTTON', () => {
			const mockPushToHistory = jest.fn()
			const mockEvent = { target: { parentElement: { className: 'checkbox' } } }
			wrapper.instance().openNote(mockEvent)
			expect(mockPushToHistory).toHaveBeenCalledTimes(0)
		})

	})

	describe.skip('onDragOver', () => {
	    it('should call props.changeNoteOrder with 2 arguments if the ID does not match', () => {
	        const mockStartId = 9
	        const mockId = 9
	        wrapper.instance().onDragOver(mockId)
	        expect(wrapper.instance().props.changeNoteOrder).toHaveBeenCalledWith(mockStartId, mockId)
			})
			
			it('should do nothing if the id matched the props.startID', () => {

			})
	})

})