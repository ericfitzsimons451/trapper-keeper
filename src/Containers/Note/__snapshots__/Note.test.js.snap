// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Note should match the snapshot 1`] = `
<div
  className="note"
  draggable={true}
  onClick={[Function]}
  onDragEnd={[Function]}
  onDragOver={[Function]}
  onDragStart={[Function]}
>
  <button
    className="delete-button"
    onClick={[Function]}
  />
  <h2>
    hello
  </h2>
  <div
    className="text"
    id={4}
    key="1554832955068"
  >
    <div
      className="uncheckbox"
      onClick={[Function]}
    />
    <p>
      yo
    </p>
  </div>
  <div
    className="text"
    id={5}
    key="1554832955068"
  >
    <div
      className="uncheckbox"
      onClick={[Function]}
    />
    <p>
      hello
    </p>
  </div>
</div>
`;
